pkg/csr/csr.go

```diff
@@ -0,0 +1,350 @@
+	}
+	groups := sets.NewString(a.csr.Spec.Groups...)
+	if !groups.HasAll(nodeGroup, systemPrefix) {
+		return errors.Errorf("CSR %s does not contain required groups", a.csr.Name)
+	}
+
+	// Check usages include: digital signature, key encipherment and server auth
+	if !hasUsages(a.csr, kubeletServerUsages) {
+		return errors.Errorf("CSR %s does not contain required usages", a.csr.Name)
+	}
+
+	var hasOrg bool
+	for i := range parsedCsr.Subject.Organization {
+		if parsedCsr.Subject.Organization[i] == nodeGroup {
+			hasOrg = true
+			break
+		}
+	}
+	if !hasOrg {
+		return errors.Errorf("CSR %s does not contain required subject organization", a.csr.Name)
+	}
+	return nil
+}
+
+// isNodeClientCert returns true if the CSR is from a  kube-apiserver-client-kubelet signer
+// reference: https://kubernetes.io/docs/reference/access-authn-authz/certificate-signing-requests/#kubernetes-signers
+func (a *Approver) isNodeClientCert(x509cr *x509.CertificateRequest) bool {
+	if !reflect.DeepEqual([]string{nodeGroup}, x509cr.Subject.Organization) {
+		return false
+	}
+	if (len(x509cr.DNSNames) > 0) || (len(x509cr.EmailAddresses) > 0) || (len(x509cr.IPAddresses) > 0) {
+		return false
+	}
+	// Check usages include: digital signature, key encipherment and client auth
+	if !hasUsages(a.csr, kubeletClientUsages) {
+		return false
+	}
+	return true
+}
+
+// hasUsages verifies if the required usages exist in the CSR spec
+func hasUsages(csr *certificates.CertificateSigningRequest, usages []certificates.KeyUsage) bool {
+	if csr == nil {
+		return false
+	}
+	if len(usages) != len(csr.Spec.Usages) {
+		return false
+	}
+	usageMap := map[certificates.KeyUsage]struct{}{}
+	for _, u := range usages {
+		usageMap[u] = struct{}{}
+	}
+
+	for _, u := range csr.Spec.Usages {
+		if _, ok := usageMap[u]; !ok {
+			return false
+		}
+	}
+	return true
+}
+
+// hasAssociatedHost returns true if the node name matches with the host name of the instances present in the configMap
+func hasAssociatedHost(nodeName string, windowsInstances []*instances.InstanceInfo,
+	instanceSigner ssh.Signer) (bool, error) {
+	for _, instance := range windowsInstances {
+		hostName, err := findHostName(instance, instanceSigner)
+		if err != nil {
+			return false, errors.Wrapf(err, "unable to find host name for instance with address %s",
+				instance.Address)
+		}
+		// check if the instance host name matches node name
+		if strings.Contains(hostName, nodeName) {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+// findHostName returns the actual host name of the instance by running the 'hostname' command
+func findHostName(instance *instances.InstanceInfo, instanceSigner ssh.Signer) (string, error) {
+	win, err := windows.New("", "", instance, instanceSigner)
+	if err != nil {
+		return "", errors.Wrap(err, "error instantiating Windows instance")
+	}
+	// get the instance host name  by running hostname command on remote VM
+	hostName, err := win.Run("hostname", true)
```

Comment:
Is `hostname` a powershell command?

Reply:
tried to answer this earlier in one of the [discussions](https://github.com/openshift/windows-machine-config-operator/pull/499#discussion_r673292797)